sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/CalculatedBackoffTime.java

```diff
@@ -0,0 +1,91 @@
+
+/**
+ * Calculates the amount of time to the next refresh, if a refresh fails.
+ */
+public final class CalculatedBackoffTime {
+
+    /**
+     * Calculates the amount of time to the next refresh, if a refresh fails. Takes current Refresh date into account
+     * for watch keys.
+     * @param state state being checked
+     * @param properties App Configuration Provider Properties
+     * @return new Refresh Date
+     */
+    public static Instant calculateBefore(State state, AppConfigurationProviderProperties properties) {
+        // The refresh interval is only updated if it is expired.
+        if (Instant.now().isAfter(state.getNextRefreshCheck())) {
+            Integer attempt = state.getRefreshAttempt();
+            state.addRefreshAttempt();
+
+            return calculate(attempt, state.getRefreshInterval(), properties);
+        }
+
+        return state.getNextRefreshCheck();
+    }
+
+    /**
+     * Calculates the amount of time to the next refresh, if a refresh fails. Takes current Refresh date into account
+     * for watch keys. Used for checking client refresh-interval only.
+     * @param nextRefreshCheck next refresh for the whole client
+     * @param attempt refresh attempt for the client
+     * @param interval the Refresh Interval
+     * @param properties App Configuration Provider Properties
+     * @return new Refresh Date
+     */
+    public static Instant calculateBefore(Instant nextRefreshCheck, Integer attempt, Duration interval,
+        AppConfigurationProviderProperties properties) {
+
+        if (interval == null) {
+            return null;
+        }
+
+        // The refresh interval is only updated if it is expired.
+        if (Instant.now().isAfter(nextRefreshCheck)) {
+            return calculate(attempt, interval.getSeconds(), properties);
+        }
+
+        return nextRefreshCheck;
+    }
+
+    private static Instant calculate(Integer attempt, long interval, AppConfigurationProviderProperties properties) {
+        int durationPeriod = Math.toIntExact(interval);
+
+        Instant now = Instant.now();
+
+        if (durationPeriod <= properties.getDefaultMinBackoff()) {
+            return now.plusSeconds(interval);
+        }
+
+        long defaultMinBackoff = TimeUnit.NANOSECONDS.convert(properties.getDefaultMinBackoff(), TimeUnit.SECONDS);
+        Integer maxRandomValue = Math.min((attempt << 2), Integer.MAX_VALUE);
+
+        long maxBackoff = Math.min(
+            TimeUnit.NANOSECONDS.convert(interval, TimeUnit.SECONDS),
+            TimeUnit.NANOSECONDS.convert(properties.getDefaultMaxBackoff(), TimeUnit.SECONDS));
+
+        double randomValue = getRandomBackoff(1, maxRandomValue);
+
+        double calculatedBackoff = defaultMinBackoff * randomValue;
+
+        // IMPORTANT: This can overflow, so set to max if less than 0
+        if (calculatedBackoff <= 0) {
+            calculatedBackoff = maxBackoff;
+        }
```

Comment:
This is the core logic of our backoff time calculations. After giving it some thought, I feel that we should keep it consistent across our services and providers. While there are many ways to calculate randomized exponential backoff, it's easier to maintain the code and catch bugs if the core logic remains the same across services. 
We should be able to reuse our C# code here without any significant changes.

Reply:
#Resolved